---
title: "Tax4Fun: Symbiodinium-bacteria interactions and functions supporting the formation of calcifying biofilms"
author: "MNitschke"
date: "19 April 2018"
output: html_document
keep_md: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Libraries

```{r  echo=TRUE}
library(Tax4Fun)
library(vegan)
library(data.table)
library(gplots)
library(multtest)
library(tibble)
library(dplyr)
library(DESeq2)
library(phyloseq)
library(tidyr)
library(stringr)
library(ggplot2)
library(scales)
```

# Import data 

Converted the biom format file to tsv format for compatibility with Tax4Fun using the BIOM-FORMAT package. This output contains the OTUs assigned to the Silva version 123 reference database so taxa names match Kegg Organisms of pre-computed data matricies in Tax4Fun.

TSV check. Check the taxonomy levels. Tax4fun requires a very specific table layout. And a cell for every taxonomic rank, even if it is empty. Otherwise the OTU will be disregarded from the composite metagenome.

```{r}
explore <- read.table(file = 'D:/Post-Doc/Symbiolite/Microbiome/Sequencing_data/Vsearch outputs/uclust_assigned_taxonomy123/otu_table132.123test.tsv', sep = '\t')

str_count(explore$V44, ";") # multiple taxonomic levels are empty. All values should = 6

t <- separate(explore, V44, sep=";", remove = TRUE, into = c("Rank1", "Rank2","Rank3","Rank4","Rank5","Rank6", "Rank7"))

t[is.na(t)]="" # Empty every taxonomic rank that has an NA in it

t$taxonomy <- paste(t$Rank1,t$Rank2,t$Rank3,t$Rank4,t$Rank5,t$Rank6, sep=";")

# Cleanup

t$Rank1 <- NULL
t$Rank2 <- NULL
t$Rank3 <- NULL
t$Rank4 <- NULL
t$Rank5 <- NULL
t$Rank6 <- NULL
t$Rank7 <- NULL

t$taxonomy <- paste(t$taxonomy, ";", sep = "")
t <- data.frame(t, stringsAsFactors = FALSE)
t[1,44] <- ""
t[2,44] <- "taxonomy"
t[1,2:44] <- ""

str_count(t$taxonomy, ";") # Now check all taxonomic levels are occupied by a string. Should contain 6 x ";" in every cell

write.table(t,"Tax4Fun_OTU.tsv",sep="\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

# Inference using Tax4Fun

Use Tax4Fun to predict the functional and metabolic capabilities of microbial communities based on 16S data samples. Suppress the option to normalise by copy number due to recommendation by Louca et al (2018) Microbiome.

```{r}
reshape <- importQIIMEData("D:/Post-Doc/Symbiolite/Microbiome/Sequencing_data/Vsearch outputs/uclust_assigned_taxonomy123/Tax4Fun_OTU.tsv")
directory <- "D:/Post-Doc/Symbiolite/Microbiome/Sequencing_data/Vsearch outputs/uclust_assigned_taxonomy123/SILVA123/"

KO <- Tax4Fun(reshape, directory, normCopyNo = FALSE)

Tax4FunProfile <- KO$Tax4FunProfile
Tax4FunProfile <- data.frame(t(KO$Tax4FunProfile))

path <- Tax4Fun(reshape, directory, normCopyNo = FALSE, fctProfiling = FALSE)

Tax4FunPath <- path$Tax4FunProfile
Tax4FunPath <- data.frame(t(path$Tax4FunProfile))
```

# Supplementary table 

Write to csv for supplementary tables in manuscript

```{r}
#write.table(Tax4FunProfile,"Tax4FunProfile_Export.csv",sep="\t")
#write.table(Tax4FunPath,"Tax4FunPath_Export.csv",sep="\t")
```

# FTUs

FTUs are the Fraction of OTUs that could not be mapped to KEGG organisms.

```{r}
path$FTU
```

# Supplementary table Tax4Fun not matched

Find which OTUs are not in the Kegg organism Tax4Fun database. 
Grab the Tax4fun hidden function using getAnywhere(importTax4FunReferenceData)

```{r}
importTax4FunReferenceData <- function (folder) 
{
    if (substr(folder, nchar(folder), nchar(folder)) == "/") {
        pathReferenceData <- folder
    }
    else {
        pathReferenceData <- paste(folder, "/", sep = "")
    }
    referenceData <- list()
    tmpReferenceData <- readRDS(paste(pathReferenceData, "KEGGBacArchTaxInformationMoPPro.RData", 
        sep = ""))
    referenceData$KEGGBacArchTaxInformationMoPPro <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "PathwayAbundancesKEGGBacArchMoPPro.RData", 
        sep = ""))
    referenceData$PathwayAbundancesKEGGBacArchMoPPro <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "PathwayInformationKEGGBacArchMoPPro.RData", 
        sep = ""))
    referenceData$PathwayInformationKEGGBacArchMoPPro <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "PathwayAbundancesKEGGBacArch.RData", 
        sep = ""))
    referenceData$PathwayAbundancesKEGGBacArch <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "PathwayInformationKEGGBacArch.RData", 
        sep = ""))
    referenceData$PathwayInformationKEGGBacArch <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "KEGGKOInformation.RData", 
        sep = ""))
    referenceData$KEGGKOInformation <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "KEGGBacArchTaxInformation.RData", 
        sep = ""))
    referenceData$KEGGBacArchTaxInformation <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "KEGGBacArchCopyNumbers.RData", 
        sep = ""))
    referenceData$KEGGBacArchCopyNumbers <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "FctAbundancesKEGGBacArchPAUDALong.RData", 
        sep = ""))
    referenceData$FctAbundancesKEGGBacArchPAUDALong <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "FctAbundancesKEGGBacArchPAUDAShort.RData", 
        sep = ""))
    referenceData$FctAbundancesKEGGBacArchPAUDAShort <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "FctAbundancesKEGGBacArchUProCLong.RData", 
        sep = ""))
    referenceData$FctAbundancesKEGGBacArchUProCLong <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "FctAbundancesKEGGBacArchUProCShort.RData", 
        sep = ""))
    referenceData$FctAbundancesKEGGBacArchUProCShort <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "SilvaToKEGGMappingMat.RData", 
        sep = ""))
    referenceData$SilvaToKEGGMappingMat <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "SilvaIDs.RData", 
        sep = ""))
    referenceData$SilvaIDs <- tmpReferenceData
    return(referenceData)
}

Tax4FunReferenceData <- importTax4FunReferenceData(directory)
commonOTUs <- intersect(Tax4FunReferenceData$SilvaIDs$V1, row.names(reshape$otuTable))
full <- row.names(reshape$otuTable)

Notpredicted <- setdiff(full, commonOTUs) # List of taxa that will not contribute to metagenomic inference
```

# Not predicted taxa percentages

Check which taxa are abundant and also not predicted.

```{r}
new <- reshape$otuTable
total <- sum(new)
new$Sum <- rowSums(new)
comdat <- data.frame(Notpredicted)
new <- tibble::rownames_to_column(new)
new <- setnames(new, 1:1, c("Notpredicted"))
NP <- left_join(comdat, new, by = "Notpredicted")
vals <- (NP$Sum/total)*100
names <- NP$Notpredicted
percentmissing <- data.frame(names, vals)
```

# Covert Paths and KO profiles to phyloseq objects

Paths first

```{r}
tab <- rownames_to_column(Tax4FunPath, var = "OTU")
tab2 <- separate(data = tab, col = OTU, into = c("OTUID", "Rank1"), sep = ";")

OTUID <- tab2$OTUID
Rank1 <- tab2$Rank1
tax <- data.frame(OTUID, Rank1)

taxonomy <- tax_table(tax)
rownames(taxonomy) <- taxonomy[,1]
taxonomy <- taxonomy[,-1]
colnames(taxonomy)[1] <- "path"

map <- read.table("D:/Post-Doc/Symbiolite/Microbiome/Sequencing_data/Vsearch outputs/uclust_assigned_taxonomy123/map.txt", header = TRUE)

sampdat <- sample_data(map)
rownames(sampdat) <- sampdat$sample
sampdat$sample <- NULL

tab3 <- tab2
tab3$Rank1 <- NULL
rownames(tab3) <- tab3[,1]
tab3[,1] <- NULL
OTUtab <- otu_table(tab3, taxa_are_rows = TRUE)

phylop <- merge_phyloseq(OTUtab, sampdat, taxonomy)
```

KO's next

```{r}
tab <- rownames_to_column(Tax4FunProfile, var = "OTU")
tab2 <- separate(data = tab, col = OTU, into = c("OTUID", "Rank1"), sep = ";")

OTUID <- tab2$OTUID
Rank1 <- tab2$Rank1
tax <- data.frame(OTUID, Rank1)

taxonomy <- tax_table(tax)
rownames(taxonomy) <- taxonomy[,1]
taxonomy <- taxonomy[,-1]
colnames(taxonomy)[1] <- "KO"

map <- read.table("D:/Post-Doc/Symbiolite/Microbiome/Sequencing_data/Vsearch outputs/uclust_assigned_taxonomy123/map.txt", header = TRUE)

sampdat <- sample_data(map)
rownames(sampdat) <- sampdat$sample
sampdat$sample <- NULL


tab3 <- tab2
tab3$Rank1 <- NULL
rownames(tab3) <- tab3[,1]
tab3[,1] <- NULL
OTUtab <- otu_table(tab3, taxa_are_rows = TRUE)

phyloKO <- merge_phyloseq(OTUtab, sampdat, taxonomy)
```

# SP vs NP

Subset the dataset by removing the temporal pre-calcifying and post-calcifying data. 

```{r}
svnp <- subset_samples(phylop, time=="n")
svnko <- subset_samples(phyloKO, time=="n")
```

# Figure 3 part 1: Ordinate

SP vs NP ordination using a Bray-Curtis matrix and PCoA ordination.

```{r}
brayp <- ordinate(svnp, "PCoA", "bray")
plot_ordination(svnp, brayp, color="state", shape="state")  + geom_text(aes(label = sample_names(svnp)), vjust = 1.5) + geom_point(size = 4) #+ scale_x_reverse()
```

# Figure 3 part 2

Grab centroids and vectors from samples to centroids to overlay on part 1

```{r}
bray <- phyloseq::distance(svnp, method = "bray")

sampledf <- data.frame(sample_data(svnp))

beta <- betadisper(bray, sampledf$state)

plot(beta)
```

# Path PERMANOVA and betadisper

Run script chunk above first and then compute statistics and homogeneity of dispersions test

```{r}
adonis(bray ~ state, data = sampledf)

permutest(beta)
```

# Path differential abundance using MT

```{r}
PMT <- mt(svnp, classlabel = "state", minPmaxT = "maxT")
alpha = 0.05
sigpath = PMT[which(PMT$adjp < alpha), ]
sigpath <- tibble::rownames_to_column(sigpath)
sigpath <- setnames(sigpath, 1:1, c("KO"))
View(sigpath)
```

# Calculate % completeness of each pathway.

Some non-elegant code for calculating the percentage of the reference-pathway completeness. Runs quickly but could be optimised to loop through pathko columns. All KOs of each pathway were retrieved from the reference pathway page at https://www.genome.jp/dbget-bin/www_bget?XXXXXXX

```{r}
pathko <- read.table(file = 'D:/Post-Doc/Symbiolite/Microbiome/Sequencing_data/Vsearch outputs/uclust_assigned_taxonomy123/T4F_path_KOs.txt', sep = '\t', header = TRUE, strip.white = TRUE)

dfko <- psmelt(svnko)

temp <- dfko[dfko$OTU %in% pathko$ko00561,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko00561 != ""))
  o <- nrow(l)/2
  ko00561 <- (o/e)*100

temp <- dfko[dfko$OTU %in% pathko$ko00564,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko00564 != ""))
  o <- nrow(l)/2
  ko00564 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko04972,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko04972 != ""))
  o <- nrow(l)/2
  ko04972 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko05111,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko05111 != ""))
  o <- nrow(l)/2
  ko05111 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko02030,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko02030 != ""))
  o <- nrow(l)/2
  ko02030 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko02060,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko02060 != ""))
  o <- nrow(l)/2
  ko02060 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko04080,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko02060 != ""))
  o <- nrow(l)/2
  ko04080 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko01054,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko01054 != ""))
  o <- nrow(l)/2
  ko01054 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko00901,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko00901 != ""))
  o <- nrow(l)/2
  ko00901 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko00300,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko00300 != ""))
  o <- nrow(l)/2
  ko00300 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko00360,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko00360 != ""))
  o <- nrow(l)/2
  ko00360 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko00906,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko00906 != ""))
  o <- nrow(l)/2
  ko00906 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko00190,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko00190 != ""))
  o <- nrow(l)/2
  ko00190 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko04961,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko04961 != ""))
  o <- nrow(l)/2
  ko04961 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko01053,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko01053 != ""))
  o <- nrow(l)/2
  ko01053 <- (o/e)*100
  
temp <- dfko[dfko$OTU %in% pathko$ko04115,]
sel <- temp %>% select(OTU, Abundance, state, KO)
  l <- sel %>% group_by(OTU,KO,state) %>% summarize(abund = mean(Abundance))
  e <- length(which(pathko$ko04115 != ""))
  o <- nrow(l)/2
  ko04115 <- (o/e)*100

KO <- colnames(pathko)
completeness <- c(ko00561,ko00564,ko04972,ko05111,ko02030,ko02060,ko04080,ko01054,ko00901,ko00300,ko00360,ko00906,ko00190,ko04961,ko01053,ko04115)
PC <- data.frame(KO, completeness)
PC <- merge(PC, sigpath, by = c("KO"))
View(PC)
```

# Figure 3 part 3

Plot pathways that are significantly different across SP vs NP strains and are of relevance to the study.

```{r}
df <- psmelt(svnp)

avgs <- df %>% select(OTU, Abundance, state, path)

d = avgs %>%
    group_by(OTU,path,state) %>%
    summarize(abund = mean(Abundance*100), sd = sd(Abundance*100))

d = d %>%
  mutate(se = ifelse(state=="calcifying", sd/sqrt(17),sd/sqrt(7)))

d$path <- as.factor(d$path)
d$state <- as.factor(d$state)

d <- d %>% filter(str_detect(path, "Glycerolipid metabolism|Bacterial chemotaxis|Vibrio cholerae pathogenic cycle|Phosphotransferase|siderophore|Glycerophospholipid|Phenylalanine metabolism|Carotenoid"))

dp <- ggplot(data=d, aes(x=state, y=abund)) + 
  geom_bar(stat = "identity") + 
  facet_grid(path~., scales = "free_y") +
  geom_errorbar(aes(ymin=abund-se, ymax=abund+se), width=0.1) + 
  scale_y_continuous() + 
  theme(aspect.ratio = 1)

dp
```

# Supplementary plot

Plot profiles to contain orthologs from Glycerolipid metabolism pathway (ko00561).

```{r}
biofilmko <- read.table(file = 'D:/Post-Doc/Symbiolite/Microbiome/Sequencing_data/Vsearch outputs/uclust_assigned_taxonomy123/T4F_path_KOs.txt', sep = '\t', header = TRUE)

dfko <- psmelt(svnko)

biofilmko$ko00561 <- as.character(biofilmko$ko00561)
dfko$OTU <- as.character(dfko$OTU)

dfko <- dfko[dfko$OTU %in% biofilmko$ko00561,]

dfkosel <- dfko %>% select(OTU, Abundance, state, KO)

d = dfkosel %>%
    group_by(OTU,KO,state) %>%
    summarize(abund = mean(Abundance), sd = sd(Abundance))

d = d %>%
  mutate(se = ifelse(state=="calcifying", sd/sqrt(17),sd/sqrt(7)))

d$KO <- as.factor(d$KO)
d$state <- as.factor(d$state)

theme_set(theme_bw())

dp <- ggplot(data=d, aes(x=KO, y=abund, colour = state)) + 
  geom_point(size = 4) + 
  geom_errorbar(aes(ymin=abund-se, ymax=abund+se)) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))# +
  scale_y_continuous(trans='log2')

dp
```

Plot profiles to contain orthologs from Biofilm pathway (ko05111, Vibrio model pathway).

```{r}
dfko <- psmelt(svnko)

biofilmko$ko05111 <- as.character(biofilmko$ko05111)
dfko$OTU <- as.character(dfko$OTU)

dfko <- dfko[dfko$OTU %in% biofilmko$ko05111,]

dfkosel <- dfko %>% select(OTU, Abundance, state, KO)

d = dfkosel %>%
    group_by(OTU,KO,state) %>%
    summarize(abund = mean(Abundance), sd = sd(Abundance))

d = d %>%
  mutate(se = ifelse(state=="calcifying", sd/sqrt(17),sd/sqrt(7)))

d$KO <- as.factor(d$KO)
d$state <- as.factor(d$state)

theme_set(theme_bw())

dp <- ggplot(data=d, aes(x=KO, y=abund, colour = state)) + 
  geom_point(size = 4) + 
  geom_errorbar(aes(ymin=abund-se, ymax=abund+se)) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  scale_y_continuous(trans='log2')

dp
```

Plot profiles to contain orthologs from Bacterial chemotaxis pathway (ko02030).

```{r}
dfko <- psmelt(svnko)

biofilmko$ko02030 <- as.character(biofilmko$ko02030)
dfko$OTU <- as.character(dfko$OTU)

dfko <- dfko[dfko$OTU %in% biofilmko$ko02030,]

dfkosel <- dfko %>% select(OTU, Abundance, state, KO)

d = dfkosel %>%
    group_by(OTU,KO,state) %>%
    summarize(abund = mean(Abundance), sd = sd(Abundance))

d = d %>%
  mutate(se = ifelse(state=="calcifying", sd/sqrt(17),sd/sqrt(7)))

d$KO <- as.factor(d$KO)
d$state <- as.factor(d$state)

theme_set(theme_bw())

dp <- ggplot(data=d, aes(x=KO, y=abund, colour = state)) + 
  geom_point(size = 4) + 
  geom_errorbar(aes(ymin=abund-se, ymax=abund+se)) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  scale_y_continuous(trans='log2')

dp
```

Plot profiles that contain orthologs from the Phosphotransferase system (PTS) pathway (ko02060).

```{r}
dfko <- psmelt(svnko)

biofilmko$ko02060 <- as.character(biofilmko$ko02060)
dfko$OTU <- as.character(dfko$OTU)

dfko <- dfko[dfko$OTU %in% biofilmko$ko02060,]

dfkosel <- dfko %>% select(OTU, Abundance, state, KO)

d = dfkosel %>%
    group_by(OTU,KO,state) %>%
    summarize(abund = mean(Abundance), sd = sd(Abundance))

d = d %>%
  mutate(se = ifelse(state=="calcifying", sd/sqrt(17),sd/sqrt(7)))

d$KO <- as.factor(d$KO)
d$state <- as.factor(d$state)

theme_set(theme_bw())

dp <- ggplot(data=d, aes(x=KO, y=abund, colour = state)) + 
  geom_point(size = 4) + 
  geom_errorbar(aes(ymin=abund-se, ymax=abund+se)) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  scale_y_continuous(trans='log2')

dp
```

# Supplementary KO ordination

```{r}
brayko <- ordinate(svnko, "PCoA", "bray")
plot_ordination(svnko, brayko, color="state", shape="state")  + geom_text(aes(label = sample_names(svnp)), vjust = 1.5) + geom_point(size = 4)# + scale_y_reverse()
```

# Supplementary KO betadisper

Run script chunk above first and then compute statistics and homogeneity of dispersions test

```{r}
bko <- phyloseq::distance(svnko, method = "bray")

sampledf <- data.frame(sample_data(svnko))

betako <- betadisper(bko, sampledf$state)

plot(betako)

```

# Supplmentary PERMANOVA and perm test of beta dispersion

```{r}
adonis(bko ~ state, data = sampledf)

permutest(betako)
```

# Supplementary KO differential abundance testing

```{r}
KMTT <- mt(svnko, classlabel = "state", minPmaxT = "maxT")
alpha = 0.05
sigMTKO = KMTT[which(KMTT$adjp < alpha), ]
View(sigMTKO)
```

# PHASE TRANSITION

Subset to only the data that has pre, active, and post calcifying phases (185, Mf, 362). 

Note: In the sample metadata culture "a" = strain 185. "b" = strain 362. "c" = strain Mfav.

```{r}
pathtemp <- subset_samples(phylop, series=="y")
KOtemp <- subset_samples(phyloKO, series=="y")
```

# Figure 4 part 1

Ordination of phases according to KEGG pathways

```{r}
bpatht <- ordinate(pathtemp, "PCoA", "bray")

plot_ordination(pathtemp, bpatht, color="culture", shape="phase")  + 
  geom_text(aes(label = sample_names(pathtemp)), vjust = 1.5) + 
  geom_point(size = 4)
```

# Figure 4 part 2 

Grab centroids and vectors from samples to centroids to overlay on part 1

```{r}
bpathtr <- phyloseq::distance(pathtemp, method = "bray")

df <- data.frame(sample_data(pathtemp))

beta2 <- betadisper(bpathtr, df$phase)

plot(beta2)
```

# PERMANOVA and betadisper

Constrain permutations to be nested within each culture to detect differences in phase.

```{r echo=TRUE, tidy=TRUE, tidy.opts=list(width.cutoff=60)}
adonis(formula = bpathtr ~ phase, data = df)

adonis(formula = bpathtr ~ culture/phase, data = df, strata = df$culture)

permutest(beta2) 
```

# Path differential abundance using MT

```{r}
PMTP <- mt(pathtemp, classlabel = "phase", minPmaxT = "maxT", test = "f")
alpha = 0.05
sigpathp = PMTP[which(PMTP$adjp < alpha), ]
sigpathp <- tibble::rownames_to_column(sigpathp)
sigpathp <- setnames(sigpathp, 1:1, c("pathway"))
View(sigpathp)
```

# Figure 4 part 3

Plot pathways that are significantly different through time and are of relevance to the study.

```{r}
positions <- c("pre", "active", "post")

df <- psmelt(pathtemp)

avgs <- df %>% select(OTU, Sample, culture, Abundance, path, phase)

d = avgs %>%
    group_by(Sample,culture,path,phase, OTU) %>%
    filter(Abundance > 0) %>%
    summarize(path.abund = sum(Abundance))

d2 = d %>%
    group_by(OTU,path,phase) %>%
    summarize(abund = mean(path.abund), se = sd(path.abund)/sqrt(3))

d2$path <- as.factor(d2$path)
d2$phase <- as.factor(d2$phase)

d3 <- d2

pick_table <- left_join(sigpathp, d3, by = "path") # Only plot the (relevant) paths that have a significant change

d3 <- d3 %>% filter(str_detect(path, "Glycerolipid metabolism|Bacterial chemotaxis|Vibrio cholerae pathogenic cycle|Phosphotransferase|siderophore|Bacterial secretion|Starch|Flagellar"))

theme_set(theme_bw())

dp <- ggplot(data=d3, aes(x=phase, y=abund*100)) + 
  geom_bar(stat = "identity") + 
  facet_grid(path~., scales = "free_y") + 
  geom_errorbar(aes(ymin=abund*100-se*100, ymax=abund*100+se*100), width=0.1) + 
  scale_x_discrete(limits = positions) + scale_y_continuous() + 
  theme(aspect.ratio = 1)

dp
```

# Supplmentary significant pathway plots

```{r}
ggplot(data=d2, aes(x=path, y=abund, colour = phase)) +
  geom_point(size = 3) +
  geom_errorbar(aes(ymin=abund-se, ymax=abund+se)) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  scale_y_continuous(trans = "log")
```

# Supplmentary KO differential abundance using MT

```{r}
PMTKO <- mt(KOtemp, classlabel = "phase", minPmaxT = "maxT", test = "f")
alpha = 0.05
sigKOp = PMTKO[which(PMTKO$adjp < alpha), ]
sigKOp <- tibble::rownames_to_column(sigKOp)
sigKOp <- setnames(sigKOp, 1:1, c("pathway"))
View(sigKOp)
```
