---
title: "Tax4Fun"
author: "MNitschke"
date: "100819"
output: html_document
keep_md: TRUE
---

# Libraries

```{r}
library(Tax4Fun)
library(vegan)
library(data.table)
library(gplots)
library(multtest)
library(tibble)
library(phyloseq)
library(tidyr)
library(stringr)
library(ggplot2)
library(scales)
library(dplyr)
```

# Import data 

Converted the biom format file to tsv format for compatibility with Tax4Fun using the BIOM-FORMAT package. This output contains the OTUs assigned to the Silva version 123 reference database so taxa names match Kegg Organisms of pre-computed data matricies in Tax4Fun.

Check otu_table132.123.tsv formatting!!! Tax4fun requires a very specific table layout that is not documented in the publication. If you convert your .biom formated OTU table to .tsv using the BIOM format tool using something like:

biom convert -i otu_table.biom -o otu_table.tsv --to-tsv --header-key taxonomy


...then the cell delimiters (likely ";") for OTUs that have not been assigned to the species level (likely to be the majority of OTUs) will not be parsed by Tax4Fun. Each taxonomy must have exactly 6 x ; delimiters, otherwise the OTU taxonomy will not be detected during the matching of SILVA IDs to KEGG organisms. Below is some code to explore a .tsv otu table and add the correct number of delimiters.

```{r}
explore <- read.table(file = 'otu_table132.123.tsv', sep = '\t')

str_count(explore$V44, ";") # multiple taxonomic levels are empty. All values should = 6

t <- separate(explore, V44, sep=";", remove = TRUE, into = c("Rank1", "Rank2","Rank3","Rank4","Rank5","Rank6", "Rank7"))

t[is.na(t)]="" # Empty every taxonomic rank that has an NA in it

t$taxonomy <- paste(t$Rank1,t$Rank2,t$Rank3,t$Rank4,t$Rank5,t$Rank6, sep=";")

# Cleanup

t$Rank1 <- NULL
t$Rank2 <- NULL
t$Rank3 <- NULL
t$Rank4 <- NULL
t$Rank5 <- NULL
t$Rank6 <- NULL
t$Rank7 <- NULL

t$taxonomy <- paste(t$taxonomy, ";", sep = "")
t <- data.frame(t, stringsAsFactors = FALSE)
t[1,44] <- ""
t[2,44] <- "taxonomy"
t[1,2:44] <- ""

str_count(t$taxonomy, ";") # Now check all taxonomic levels are occupied by a string. Should contain 6 x ";" in every cell

write.table(t,"Tax4Fun_OTU.tsv", sep="\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

# Inference using Tax4Fun

Use Tax4Fun to predict the functional and metabolic capabilities of microbial communities based on 16S data samples

```{r}
reshape <- importQIIMEData("Tax4Fun_OTU.tsv")
directory <- "D:/Post-Doc/Symbiolite/Microbiome/Sequencing_data/Vsearch outputs/uclust_assigned_taxonomy123/SILVA123/"

KO <- Tax4Fun(reshape, directory, normCopyNo = TRUE)

Tax4FunProfile <- KO$Tax4FunProfile
Tax4FunProfile <- data.frame(t(KO$Tax4FunProfile))

path <- Tax4Fun(reshape, directory, normCopyNo = TRUE, fctProfiling = FALSE)

Tax4FunPath <- path$Tax4FunProfile
Tax4FunPath <- data.frame(t(path$Tax4FunProfile))
```

## Supplementary file: Tax4Fun profiles and paths

Write to csv for supplementary tables in manuscript

```{r}
write.table(Tax4FunProfile,"Tax4FunProfile_Export.csv",sep="\t")
write.table(Tax4FunPath,"Tax4FunPath_Export.csv",sep="\t")
```

## Supplementary file: Tax4Fun FTUs

FTUs are the Fraction of OTUs that could not be mapped to KEGG organisms.

```{r}
path$FTU
```

## Supplementary file: Taxa (otus) not predicted by Tax4Fun

Find which OTUs are not in the Kegg organism Tax4Fun database (for some reason Tax4Fun does not provide this as output) 
Retrieve the Tax4fun hidden function using getAnywhere(importTax4FunReferenceData)

```{r}
importTax4FunReferenceData <- function (folder) 
{
    if (substr(folder, nchar(folder), nchar(folder)) == "/") {
        pathReferenceData <- folder
    }
    else {
        pathReferenceData <- paste(folder, "/", sep = "")
    }
    referenceData <- list()
    tmpReferenceData <- readRDS(paste(pathReferenceData, "KEGGBacArchTaxInformationMoPPro.RData", 
        sep = ""))
    referenceData$KEGGBacArchTaxInformationMoPPro <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "PathwayAbundancesKEGGBacArchMoPPro.RData", 
        sep = ""))
    referenceData$PathwayAbundancesKEGGBacArchMoPPro <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "PathwayInformationKEGGBacArchMoPPro.RData", 
        sep = ""))
    referenceData$PathwayInformationKEGGBacArchMoPPro <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "PathwayAbundancesKEGGBacArch.RData", 
        sep = ""))
    referenceData$PathwayAbundancesKEGGBacArch <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "PathwayInformationKEGGBacArch.RData", 
        sep = ""))
    referenceData$PathwayInformationKEGGBacArch <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "KEGGKOInformation.RData", 
        sep = ""))
    referenceData$KEGGKOInformation <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "KEGGBacArchTaxInformation.RData", 
        sep = ""))
    referenceData$KEGGBacArchTaxInformation <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "KEGGBacArchCopyNumbers.RData", 
        sep = ""))
    referenceData$KEGGBacArchCopyNumbers <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "FctAbundancesKEGGBacArchPAUDALong.RData", 
        sep = ""))
    referenceData$FctAbundancesKEGGBacArchPAUDALong <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "FctAbundancesKEGGBacArchPAUDAShort.RData", 
        sep = ""))
    referenceData$FctAbundancesKEGGBacArchPAUDAShort <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "FctAbundancesKEGGBacArchUProCLong.RData", 
        sep = ""))
    referenceData$FctAbundancesKEGGBacArchUProCLong <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "FctAbundancesKEGGBacArchUProCShort.RData", 
        sep = ""))
    referenceData$FctAbundancesKEGGBacArchUProCShort <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "SilvaToKEGGMappingMat.RData", 
        sep = ""))
    referenceData$SilvaToKEGGMappingMat <- tmpReferenceData
    tmpReferenceData <- readRDS(paste(pathReferenceData, "SilvaIDs.RData", 
        sep = ""))
    referenceData$SilvaIDs <- tmpReferenceData
    return(referenceData)
}

Tax4FunReferenceData <- importTax4FunReferenceData(directory)
commonOTUs <- intersect(Tax4FunReferenceData$SilvaIDs$V1, row.names(reshape$otuTable))
full <- row.names(reshape$otuTable)

Notpredicted <- setdiff(full, commonOTUs) # List of taxa that will not contribute to metagenomic inference
```

# Covert Paths and KO profiles to phyloseq objects

## Paths first

```{r}
tab <- rownames_to_column(Tax4FunPath, var = "OTU")
tab2 <- separate(data = tab, col = OTU, into = c("OTUID", "Rank1"), sep = ";")

OTUID <- tab2$OTUID
Rank1 <- tab2$Rank1
tax <- data.frame(OTUID, Rank1)

taxonomy <- tax_table(tax)
rownames(taxonomy) <- taxonomy[,1]
taxonomy <- taxonomy[,-1]
colnames(taxonomy)[1] <- "path"

map <- read.table("map.txt", header = TRUE)

sampdat <- sample_data(map)
rownames(sampdat) <- sampdat$sample
sampdat$sample <- NULL

tab3 <- tab2
tab3$Rank1 <- NULL
rownames(tab3) <- tab3[,1]
tab3[,1] <- NULL
OTUtab <- otu_table(tab3, taxa_are_rows = TRUE)

phylop <- merge_phyloseq(OTUtab, sampdat, taxonomy)
```

## KO's next

```{r}
tab <- rownames_to_column(Tax4FunProfile, var = "OTU")
tab2 <- separate(data = tab, col = OTU, into = c("OTUID", "Rank1"), sep = ";")

OTUID <- tab2$OTUID
Rank1 <- tab2$Rank1
tax <- data.frame(OTUID, Rank1)

taxonomy <- tax_table(tax)
rownames(taxonomy) <- taxonomy[,1]
taxonomy <- taxonomy[,-1]
colnames(taxonomy)[1] <- "KO"

map <- read.table("map.txt", header = TRUE)

sampdat <- sample_data(map)
rownames(sampdat) <- sampdat$sample
sampdat$sample <- NULL

tab3 <- tab2
tab3$Rank1 <- NULL
rownames(tab3) <- tab3[,1]
tab3[,1] <- NULL
OTUtab <- otu_table(tab3, taxa_are_rows = TRUE)

phyloKO <- merge_phyloseq(OTUtab, sampdat, taxonomy)
```

Add more sample metadata to the phyloseq objects about clade and ITS2 type

```{r}
ITS2 <- c("A2","A2","B1","B1","A2","A2","A2","A2","A2","A2","A2","A2","A2","B1","A1","A1","A1","A1","A1","A1","A1","A1","A1","A4","B1","A2","B1","A12","A2","A2","A2","A2","A2","A2","A2","A2","A2","B1","B1","B1","B1","B1")
clade <- c("A","A","B","B","A","A","A","A","A","A","A","A","A","B","A","A","A","A","A","A","A","A","A","A","B","A","B","A","A","A","A","A","A","A","A","A","A","B","B","B","B","B")

sample_data(phylop)$ITS2 <- ITS2
sample_data(phylop)$clade <- clade

sample_data(phyloKO)$ITS2 <- ITS2
sample_data(phyloKO)$clade <- clade
```

# SP vs NP

Subset the dataset by removing the temporal pre-calcifying and post-calcifying data. 

```{r}
svnp <- subset_samples(phylop, time == "n")
svnko <- subset_samples(phyloKO, time == "n")
```

## Figure 2a part 1: Unconstrained ordination

SP vs NP ordination using a Bray-Curtis matrix and PCoA ordination.

```{r}
brayp <- ordinate(svnp, "PCoA", "bray")
plot_ordination(svnp, brayp, color = "state", shape = "clade") + 
  geom_text(aes(label = sample_names(svnp)), vjust = 1.5) + 
  geom_point(size = 4) + 
  theme(aspect.ratio = 1)
```

## Figure 2a part 2

Grab centroids and vectors from samples to centroids to overlay on part 1

```{r}
bray <- phyloseq::distance(svnp, method = "bray")

sampledf <- data.frame(sample_data(svnp))

betastate <- betadisper(bray, sampledf$state)

plot(betastate)
```

## Path PERMANOVA and betadisper

### Supplementary table S5:

Run script chunk above first and then compute statistics and homogeneity of dispersions test

```{r}
adonis2(bray ~ state + clade, by = "margin", data = sampledf)

anova(betastate)
```

## Supplementary file: Tax4Fun Path differential abundance using MT

```{r}
PMT <- mt(svnp, classlabel = "state", minPmaxT = "maxT")
alpha = 0.05
sigpath = PMT[which(PMT$adjp < alpha), ]
sigpath <- tibble::rownames_to_column(sigpath)
sigpath <- setnames(sigpath, 1:1, c("KO"))
View(sigpath)
```

## Calculate % completeness of pathways that are of potential relevance to the study.

All KOs of each pathway were retrieved from the reference pathway page at https://www.genome.jp/dbget-bin/www_bget?ko#####

```{r}
targetpaths <- c(sigpath$KO)

refpaths <- read.table(file = 'reference_path_KO.txt', sep = '\t', header = TRUE, strip.white = TRUE)
refpaths[refpaths==""] <- NA

completeness <- function(pathls, phyloseqpath, phyloseqko, refpaths){
  subls <- as.matrix(subset(refpaths, select = pathls)) # subset reference pathway file (KO's), headed by pathways (ko's) to match query list
  pkos <- as.matrix(rownames(tax_table(phyloseqko))) # get a list of all KO's returned by Tax4Fun from phyloseq object
  pkodf <- as.matrix(pkos)
  common <- apply(subls, 2, function(x) intersect(x, pkodf)) # intersect every reference pathway KO complete list against what Tax4Fun retrieved
  found <- apply(subls,2, function(x) length(which(!is.na(x)))) # get a count of orthologs for each pathway
  pc <- (lengths(common) / found) * 100 # calculate % orthologs of reference pathway present in Tax4Fun output
  ko <- pathls
  df <- data.frame(ko, pc)
  pathdf <- as.data.frame(tax_table(phyloseqpath))
  pathdf <- rownames_to_column(pathdf, var = "ko")
  output <- left_join(df, pathdf, by = "ko")
  return(output)
}

total_comp <- completeness(targetpaths, svnp, svnko, refpaths)
```

Pathway completeness SP versus NP

```{r}
svnp_SP <- subset_samples(svnp, state == "SP")
svkop_SP <- subset_samples(svnko, state == "SP")
svkop_SP <- filter_taxa(svkop_SP, function(x) sum(x) > 0, TRUE)

SP_comp <- completeness(targetpaths, svnp_SP, svkop_SP, refpaths)

svnp_NP <- subset_samples(svnp, state == "NP")
svkop_NP <- subset_samples(svnko, state == "NP")
svkop_NP <- filter_taxa(svkop_NP, function(x) sum(x) > 0, TRUE)

NP_comp <- completeness(targetpaths, svnp_NP, svkop_NP, refpaths)
```

Now compute on a sample by sample basis by looping

```{r}
sample_data(svnp)$names <- sample_names(svnp)
single_sample_p <- list()
for(i in 1:length(sample_data(svnp)$names)){
    name <- as.character(sample_data(svnp)$names[i])
    single_sample_p[[i]] <- subset_samples(svnp, names == name)
}

sample_data(svnko)$names <- sample_names(svnko)
single_sample_ko <- list()
for(i in 1:length(sample_data(svnko)$names)){
    name <- as.character(sample_data(svnko)$names[i])
    single_sample_ko[[i]] <- subset_samples(svnko, names == name)
    single_sample_ko[[i]] <- filter_taxa(single_sample_ko[[i]], function(x) sum(x) > 0, TRUE)
}

per_sample_comp_list <- list()
for(i in 1:length(single_sample_p)){
    single_comp <- completeness(targetpaths, single_sample_p[[i]], single_sample_ko[[i]], refpaths)
    per_sample_comp_list[[i]] <- single_comp
}

key <- data.frame(names = sample_names(svnp), state = sample_data(svnp)$state, id = as.character(seq(1, length(sample_names(svnp)), 1)))
per_sample_comp <- bind_rows(per_sample_comp_list, .id = "id") %>%
  left_join(., key)

per_sample_comp_summary <- per_sample_comp %>%
  group_by(ko, path, state) %>%
  summarise(mean_pc = mean(pc),
            sd_pc = sd(pc))
```

## Figure 2b

Plot pathways that are significantly different across SP vs NP strains and are of relevance to the study.

```{r}
df <- psmelt(svnp)

avgs <- df %>% dplyr::select(OTU, Abundance, state, path)

d = avgs %>%
    group_by(OTU,path,state) %>%
    summarize(abund = mean(Abundance*100), sd = sd(Abundance*100))

d = d %>%
  mutate(se = ifelse(state == "SP", sd/sqrt(17),sd/sqrt(7)))

d$path <- as.factor(d$path)
d$state <- as.factor(d$state)

d2 <- d %>% filter(str_detect(path, "Bacterial chemotaxis|Vibrio cholerae pathogenic cycle|Phosphotransferase|siderophore"))

dp <- ggplot(data=d2, aes(x=state, y=abund)) + 
  geom_bar(stat = "identity") + 
  facet_grid(path~., scales = "free_y") +
  geom_errorbar(aes(ymin=abund-se, ymax=abund+se), width=0.1) + 
  scale_y_continuous() + 
  theme(aspect.ratio = 1)

dp
```

# SP metagenomes through time

Subset to only the data that has pre, active, and post calcifying phases (185, Mf, 362). 

Note: In the sample metadata culture "a" = strain 185. "b" = strain 362. "c" = strain Mfav.

```{r}
pathtemp <- subset_samples(phylop, series == "y")
KOtemp <- subset_samples(phyloKO, series == "y")
```

## Figure 4a part 1

Ordination of phases according to KEGG pathways

```{r}
bpatht <- ordinate(pathtemp, "PCoA", "bray")

plot_ordination(pathtemp, bpatht, color = "phase", shape = "culture")  + 
  geom_text(aes(label = sample_names(pathtemp)), vjust = 1.5) + 
  geom_point(size = 4)+
  theme(aspect.ratio = 1)
```

## Figure 4a part 2 

Grab centroids and vectors from samples to centroids to overlay on part 1

```{r}
bpathtr <- phyloseq::distance(pathtemp, method = "bray")

df <- data.frame(sample_data(pathtemp))

beta2 <- betadisper(bpathtr, df$phase)

plot(beta2)
```

## PERMANOVA and betadisper

### Supplementary table S9: 

Constrain permutations to be nested within each culture to detect differences in phase.

```{r}
adonis(formula = bpathtr ~ phase, data = df)

adonis(formula = bpathtr ~ culture/phase, data = df, strata = df$culture)

anova(beta2)
```

## Supplmentary file: Tax4fun Path differential abundance through time using MT

```{r}
PMTP <- mt(pathtemp, classlabel = "phase", minPmaxT = "maxT", test = "f")
alpha = 0.05
sigpathp = PMTP[which(PMTP$adjp < alpha), ]
sigpathp <- tibble::rownames_to_column(sigpathp)
sigpathp <- setnames(sigpathp, 1:1, c("KO"))
View(sigpathp)
```