---
title: "piphillin"
author: "MNitschke"
date: "100819"
output: html_document
---

# Libraries

```{r  echo=TRUE}
library(DESeq2)
library(vegan)
library(data.table)
library(gplots)
library(multtest)
library(tibble)
library(dplyr)
library(phyloseq)
library(tidyr)
library(stringr)
library(ggplot2)
library(scales)
```

# Import Piphillin pathway data and convert to phyloseq object

```{r}
tax <- read.table("piphillin90_pathwayko.txt", header = TRUE, sep = "\t")

taxnames <- read.table("rest_path.txt", header = FALSE, colClasses=c("character", "character", "character"), sep = "\t")
taxnames$OTUID <- paste(taxnames$V1, taxnames$V2, sep="")
taxnames$V2 <- NULL
taxnames$V1 <- NULL

taxonomy <- left_join(tax, taxnames, by = "OTUID")

taxonomy <- tax_table(taxonomy)
rownames(taxonomy) <- taxonomy[,1]
taxonomy <- taxonomy[,-1]
colnames(taxonomy)[1] <- "path"

otu <- read.table("piphillin90_OTUtabpathways.txt", header = TRUE, sep = "\t")

is.num <- sapply(otu, is.numeric)
otu[is.num] <- lapply(otu[is.num], floor)

rownames(otu) <- otu[,1]
otu[,1] <- NULL
OTUtab <- otu_table(otu, taxa_are_rows = TRUE)

map <- read.table("map.txt", header = TRUE)

sampdat <- sample_data(map)
rownames(sampdat) <- sampdat$sample
sampdat$sample <- NULL

pippath <- merge_phyloseq(OTUtab, sampdat, taxonomy)
```

# Import Piphillin ortholog data and convert to phyloseq object

```{r}
tax <- read.table("piphillin90_KO.txt", header = TRUE, sep = "\t")

taxnames <- read.table("rest_ko.txt", header = FALSE, colClasses=c("character", "character"), sep = "\t", quote="")

taxnames <- separate(taxnames, V2, sep=";", remove = TRUE, into = c("name", "definition"))
colnames(taxnames)[1] <- "OTUID"

taxonomy <- left_join(tax, taxnames, by = "OTUID")

taxonomy <- tax_table(taxonomy)
rownames(taxonomy) <- taxonomy[,1]
taxonomy <- taxonomy[,-1]
colnames(taxonomy)[1] <- "name"
colnames(taxonomy)[2] <- "definition"

otu <- read.table("piphillin90_OTUtabKOs.txt", header = TRUE, sep = "\t")

is.num <- sapply(otu, is.numeric)
otu[is.num] <- lapply(otu[is.num], floor)

rownames(otu) <- otu[,1]
otu[,1] <- NULL
OTUtab <- otu_table(otu, taxa_are_rows = TRUE)

map <- read.table("map.txt", header = TRUE)

sampdat <- sample_data(map)
rownames(sampdat) <- sampdat$sample
sampdat$sample <- NULL

pipko <- merge_phyloseq(OTUtab, sampdat, taxonomy)
```

Add some more sample metadata to the phyloseq objects about clade and ITS2 type

```{r}
ITS2 <- c("A2","A2","B1","B1","A2","A2","A2","A2","A2","A2","A2","A2","A2","B1","A1","A1","A1","A1","A1","A1","A1","A1","A1","A4","B1","A2","B1","A12","A2","A2","A2","A2","A2","A2","A2","A2","A2","B1","B1","B1","B1","B1")
clade <- c("A","A","B","B","A","A","A","A","A","A","A","A","A","B","A","A","A","A","A","A","A","A","A","A","B","A","B","A","A","A","A","A","A","A","A","A","A","B","B","B","B","B")

sample_data(pippath)$ITS2 <- ITS2
sample_data(pippath)$clade <- clade

sample_data(pipko)$ITS2 <- ITS2
sample_data(pipko)$clade <- clade
```

# Supplementary file: Piphillin KEGG pathways and KEGG orthologs

Write to csv for supplementary tables in manuscript

```{r}
pathcsv <- as.data.frame((otu_table(pippath)))
pathcsv <- tibble::rownames_to_column(pathcsv)
setnames(pathcsv, 1:1, c("ko"))

taxcsv <- as.data.frame((tax_table(pippath)))
taxcsv <- tibble::rownames_to_column(taxcsv)
setnames(taxcsv, 1:1, c("ko"))

pathcsv <- left_join(pathcsv, taxcsv, by = "ko")

write.table(pathcsv,"Piphillin_path_export.csv",sep="\t")

pathcsv <- as.data.frame((otu_table(pipko)))
pathcsv <- tibble::rownames_to_column(pathcsv)
setnames(pathcsv, 1:1, c("ko"))

taxcsv <- as.data.frame((tax_table(pipko)))
taxcsv <- tibble::rownames_to_column(taxcsv)
setnames(taxcsv, 1:1, c("ko"))

pathcsv <- left_join(pathcsv, taxcsv, by = "ko")

write.table(pathcsv,"Piphillin_ko_export.csv",sep="\t")
```

# SP vs NP

Subset the dataset by removing the temporal pre-calcifying and post-calcifying data. 

```{r echo=TRUE,  tidy=TRUE, tidy.opts=list(width.cutoff=60)}
SPNPp <- subset_samples(pippath, time=="n")
SPNPko <- subset_samples(pipko, time=="n")
```

## Figure 2a part 1: Unconstrained ordinations

SP vs NP ordination using a Bray-Curtis matrix and PCoA ordination.

```{r}
SPNPt = transform_sample_counts(SPNPp, function(x) x/sum(x))

brayp <- ordinate(SPNPt, "PCoA", "bray")

plot_ordination(SPNPt, brayp, color="state", shape="clade") + 
  geom_text(aes(label = sample_names(SPNPt)), vjust = 1.5) + 
  geom_point(size = 4) + 
  theme(aspect.ratio = 1)
```

## Figure 2a part 2

Grab centroids and vectors from samples to centroids to overlay on part 1

```{r}
bray <- phyloseq::distance(SPNPt, method = "bray")
sampledf <- data.frame(sample_data(SPNPt))

betastate <- betadisper(bray, sampledf$state)

plot(betastate)
```

## Path PERMANOVA and betadisper

### Supplementary table S5

Run script chunk above first and then compute statistics and homogeneity of dispersions test

```{r}
adonis2(bray ~ state + clade, by = "margin", data = sampledf)

anova(betastate)
```

## Differential abundance

### Supplementary file: Piphillin SPvNP pathway differential abundance

Use DESeq2 to test for differential abundance of KEGG pathways between SP and NP strains.

```{r}
SPvNP <- phyloseq_to_deseq2(SPNPp, ~ state)
SPvNP = DESeq(SPvNP, test="Wald", fitType="parametric")
res = results(SPvNP, cooksCutoff = FALSE)
alpha = 0.05
sigtab = res[which(res$padj < alpha), ]
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(SPNPp)[rownames(sigtab), ], "matrix"))
View(sigtab)
```


## Figure 2b: Pathways of interest

Plot pathways that are significantly different across SP vs NP strains and are of relevance to the study.

```{r}
df <- psmelt(SPNPt)

avgs <- df %>% dplyr::select(OTU, Abundance, state, path)

d = avgs %>%
    group_by(OTU,path,state) %>%
    summarize(abund = mean(Abundance*100), sd = sd(Abundance*100))

d = d %>%
  mutate(se = ifelse(state=="SP", sd/sqrt(17),sd/sqrt(7)))

d$path <- as.factor(d$path)
d$state <- as.factor(d$state)

d2 <- d %>% filter(str_detect(path, "chemotaxis|Phosphotransferase|Biofilm formation - Ps|Flagellar"))

dp <- ggplot(data=d2, aes(x=state, y=abund)) + 
  geom_bar(stat = "identity") + 
  facet_grid(path~., scales = "free_y") +
  geom_errorbar(aes(ymin=abund-se, ymax=abund+se), width=0.1) + 
  scale_y_continuous() + 
  theme(aspect.ratio = 1)

dp
```

## Calculate % completeness of each pathway plotted above.

All KOs of each pathway were retrieved from the reference pathway page at https://www.genome.jp/dbget-bin/www_bget?ko#####

```{r}
targetpaths <- c("ko00061","ko00062","ko02060","ko02040","ko00564","ko01051","ko00906","ko04974","ko04973","ko02030","ko02026","ko05111","ko02025")

refpaths <- read.table(file = 'D:/Post-Doc/Symbiolite/Microbiome/Sequencing_data/Vsearch outputs/uclust_assigned_taxonomy123/reference_path_KO.txt', sep = '\t', header = TRUE, strip.white = TRUE)
refpaths[refpaths==""] <- NA

completeness <- function(pathls, phyloseqpath, phyloseqko, refpaths){
  subls <- as.matrix(subset(refpaths, select = pathls)) # subset reference pathway file (KO's), headed by pathways (ko's) to match query list
  pkos <- as.matrix(rownames(tax_table(phyloseqko))) # get a list of all KO's returned by piphillin from phyloseq object
  pkodf <- as.matrix(pkos)
  common <- apply(subls, 2, function(x) intersect(x, pkodf)) # intersect every reference pathway KO complete list against what piphillin retrieved
  found <- apply(subls,2, function(x) length(which(!is.na(x)))) # get a count of orthologs for each pathway
  pc <- (lengths(common) / found) * 100 # calculate % orthologs of reference pathway present in piphillin output
  ko <- pathls
  df <- data.frame(ko, pc)
  pathdf <- as.data.frame(tax_table(phyloseqpath))
  pathdf <- rownames_to_column(pathdf, var = "ko")
  output <- left_join(df, pathdf, by = "ko")
  return(output)
}

PC <- completeness(targetpaths, SPNPp, SPNPko, refpaths)

PC
```

## Contrast peptidases in SP and NP strains

### Supplementary figure S4a

Using Ko identifiers from the Kegg rest API (http://rest.kegg.jp/get/br:ko01002/), pull out the peptidases that are significantly different between SP and NP strains. Write the output sigtabko for supplementary.

```{r}
SPvNPko <- phyloseq_to_deseq2(SPNPko, ~ state)
SPvNPko = DESeq(SPvNPko, test="Wald", fitType="parametric")
res = results(SPvNPko, cooksCutoff = FALSE)
alpha = 0.05
sigtabko = res[which(res$padj < alpha), ]
sigtabko = cbind(as(sigtabko, "data.frame"), as(tax_table(SPNPko)[rownames(sigtabko), ], "matrix"))

rest <- read.table(file = 'rest_family.txt', sep = '\t', header = TRUE, strip.white = TRUE)

df <- sigtabko
df <- tibble::rownames_to_column(df, var = "ko")

colnames(rest)[1] <- "ko"

df <- merge(rest, df, by = "ko") 

ggplot(df, aes(y=reorder(definition, log2FoldChange, sum), x=log2FoldChange)) + 
                      geom_errorbarh(aes(xmin = log2FoldChange + lfcSE, xmax = log2FoldChange - lfcSE), height = 0.2, colour = "black") + 
                      geom_point(size=4, show.legend = F) +
                      geom_vline(xintercept = 0) +
                      theme(aspect.ratio = 16/9)
```

## Contrast nucleases in SP and NP strains

### Supplementary figure S4b

```{r}
sigtabp <- sigtabko %>% filter(str_detect(definition,"nuclease|dnase"))

ggplot(sigtabp, aes(y=reorder(definition, log2FoldChange, sum), x=log2FoldChange)) + 
        geom_errorbarh(aes(xmin = log2FoldChange + lfcSE, xmax = log2FoldChange - lfcSE), height = 0.2, colour = "black") + 
        geom_point(size=4, show.legend = F, colour = "black") +
        geom_vline(xintercept = 0) +
        theme(aspect.ratio = 16/9)
```

# SP metagenomes through time

```{r}
pathtemp <- subset_samples(pippath, series=="y")
kotemp <- subset_samples(pipko, series=="y")
```

## Figure 4b: Unconstrained ordination

Transform to relative proportions as above. Pre vs active vs post calcification ordinations using a Bray-Curtis matrix and PCoA ordination.

```{r echo=TRUE, tidy=TRUE, tidy.opts=list(width.cutoff=60)}
temp.t = transform_sample_counts(pathtemp, function(x) x/sum(x))

brayordtranstemp <- ordinate(temp.t, "PCoA", "bray")

plot_ordination(temp.t, brayordtranstemp, color="phase", shape="culture")  + 
  geom_text(aes(label = sample_names(temp.t)), vjust = 1.5) +
  geom_point(size = 4) +
  scale_x_reverse() +
  theme(aspect.ratio = 1)
```

## Figure 4b part 2 

Grab centroids and vectors from samples to centroids to overlay on part 1

```{r}
bpathtr <- phyloseq::distance(temp.t, method = "bray")

df <- data.frame(sample_data(temp.t))

beta2 <- betadisper(bpathtr, df$phase)

plot(beta2)
```

## PERMANOVA and betadisper

### Supplementary table S6

Constrain permutations to be nested within each culture to detect differences in phase.

```{r echo=TRUE, tidy=TRUE, tidy.opts=list(width.cutoff=60)}
adonis(formula = bpathtr ~ phase, data = df)

adonis(formula = bpathtr ~ culture/phase, data = df, strata = df$culture)

anova(beta2)
```

## Differential abundance

### Supplementary file: Piphillin pathways that are differentially abundant through time

```{r}
phaseda <- phyloseq_to_deseq2(pathtemp, ~ culture * phase) #Run this to explore across cultures

phaseda = DESeq(phaseda, test="Wald", fitType="parametric")

res.p.a = results(phaseda, cooksCutoff = FALSE, contrast = c("phase", "pre", "active"))
res.a.po = results(phaseda, cooksCutoff = FALSE, contrast = c("phase", "active", "post"))
res.p.po = results(phaseda, cooksCutoff = FALSE, contrast = c("phase", "pre", "post"))

alpha = 0.05

sigtab.p.a = res.p.a[which(res.p.a$padj < alpha), ]
sigtab.p.a = cbind(as(sigtab.p.a, "data.frame"), as(tax_table(pathtemp)[rownames(sigtab.p.a), ], "matrix"))

View(sigtab.p.a)

sigtab.a.po = res.a.po[which(res.a.po$padj < alpha), ]
sigtab.a.po = cbind(as(sigtab.a.po, "data.frame"), as(tax_table(pathtemp)[rownames(sigtab.a.po), ], "matrix"))

View(sigtab.a.po)

sigtab.p.po = res.p.po[which(res.p.po$padj < alpha), ]
sigtab.p.po = cbind(as(sigtab.p.po, "data.frame"), as(tax_table(pathtemp)[rownames(sigtab.p.po), ], "matrix"))

View(sigtab.p.po)
```